[mini] Fix check for FastAllocateString that used old name

filipnavara opened the pull request:
The check that was supposed to take managed allocation code path for strings was broken since e58097e957af9939c94ee46ecd78e38b16a1f989. It went unnoticed for over 4 years.

marek-safar commented:
This needs monolite bump

filipnavara commented:
@marek-safar I don't think it does. I didn't change the icall name or order, it was already correct.

filipnavara commented:
The test failures are related. Looking into them.

```
1) Constructor4_LargeString (MonoTests.System.StringTest.Constructor4_LargeString)
   System.OutOfMemoryException : Insufficient memory to continue the execution of the program.
  at (wrapper alloc) System.Object.AllocString(intptr,int)
  at System.String.Ctor (System.Char c, System.Int32 count) [0x0001d] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/external/corefx/src/Common/src/CoreLib/System/String.cs:302 
  at System.String.CreateString (System.Char c, System.Int32 count) [0x00000] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/ReferenceSources/String.cs:373 
  at (wrapper managed-to-managed) System.String..ctor(char,int)
  at MonoTests.System.StringTest.Constructor4_LargeString () [0x00005] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/Test/System/StringTest.cs:126 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395 


2) PadLeft_LargeString (MonoTests.System.StringTest.PadLeft_LargeString)
   System.OutOfMemoryException : Insufficient memory to continue the execution of the program.
  at (wrapper alloc) System.Object.AllocString(intptr,int)
  at System.String.PadLeft (System.Int32 totalWidth, System.Char paddingChar) [0x00025] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/external/corefx/src/Common/src/CoreLib/System/String.Manipulation.cs:827 
  at MonoTests.System.StringTest.PadLeft_LargeString () [0x00005] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/Test/System/StringTest.cs:3011 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395 


3) PadRight_LargeString (MonoTests.System.StringTest.PadRight_LargeString)
   System.OutOfMemoryException : Insufficient memory to continue the execution of the program.
  at (wrapper alloc) System.Object.AllocString(intptr,int)
  at System.String.PadRight (System.Int32 totalWidth, System.Char paddingChar) [0x00025] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/external/corefx/src/Common/src/CoreLib/System/String.Manipulation.cs:853 
  at MonoTests.System.StringTest.PadRight_LargeString () [0x00005] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/Test/System/StringTest.cs:3063 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395 
```

EgorBo commented on mono/metadata/sgen-mono-ilgen.c:
```diff
@@ -312,13 +312,17 @@ emit_managed_allocator_ilgen (MonoMethodBuilder *mb, gboolean slowpath, gboolean
+		 * 
+		 * On 64-bit platforms SIZE_MAX is so big that the 32-bit string length can
+		 * never reach the maximum size.
 		 */
```
Unrelated: it reminds me there are tests in corefx that expect `new int[int.MaxValue]` to crash because the allowed size is something like int.MaxValue - objsize something like that

filipnavara replied:
Also unrelated, the managed code generated for `MONO_BIG_ARRAYS` seems to have the size check wrong.

marek-safar commented:
>  I didn't change the icall name or order, it was already correct.

oh, I missed that

filipnavara commented:
@monojenkins build Linux AArch64 Coop Suspend

EgorBo commented:
@monojenkins build failed

filipnavara commented:
Failure in `mono-coreonly-CI (Build macOS x64)` is unrelated (https://github.com/mono/mono/issues/17224).

filipnavara commented:
Preliminary performance numbers:

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Tests.Perf_String.Substring_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 10, i2: |      8.99 |           107.84 |            12.00 |         |
| System.Tests.Perf_String.Substring_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 7, i2:  |      8.57 |           107.96 |            12.59 |         |
| System.Tests.Perf_String.Substring_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 0, i2:  |      8.43 |           109.22 |            12.96 |         |
| System.Tests.Perf_String.Remove_Int(s: "dzsdzsDDZSDZSDZSddsz", i: 7)             |      7.77 |           111.63 |            14.37 |         |
| System.Tests.Perf_String.Substring_Int(s: "dzsdzsDDZSDZSDZSddsz", i: 10)         |      7.39 |           114.61 |            15.51 |         |
| System.Tests.Perf_String.Substring_Int(s: "dzsdzsDDZSDZSDZSddsz", i: 7)          |      7.35 |           114.51 |            15.58 |         |
| System.Tests.Perf_String.Remove_Int(s: "dzsdzsDDZSDZSDZSddsz", i: 10)            |      7.23 |           113.58 |            15.71 |         |
| System.Tests.Perf_String.Remove_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 7, i2: 4)  |      6.43 |           114.17 |            17.76 |         |
| System.Tests.Perf_String.Remove_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 0, i2: 8)  |      6.36 |           112.68 |            17.72 |         |
| System.Tests.Perf_String.Remove_IntInt(s: "dzsdzsDDZSDZSDZSddsz", i1: 10, i2: 1) |      6.34 |           120.25 |            18.96 |         |
| System.Tests.Perf_String.Split(s: "A B C D E F G H I J K L M N O P Q R S T U V W |      6.09 |          3062.62 |           502.68 |         |
| System.Tests.Perf_String.TrimStart_CharArr(s: " Test", c: [' ', ' '])            |      6.02 |           116.02 |            19.27 |         |
| System.Tests.Perf_String.TrimEnd_CharArr(s: "Test ", c: [' ', ' '])              |      5.98 |           118.18 |            19.75 |         |
| System.Tests.Perf_String.Split(s: "A B C D E F G H I J K L M N O P Q R S T U V W |      5.88 |          3096.21 |           526.37 |         |
| System.Tests.Perf_String.Trim_CharArr(s: " Test", c: [' ', ' '])                 |      5.48 |           120.00 |            21.91 |         |
| System.Tests.Perf_String.Insert(s1: "Test", i: 2, s2: " Test")                   |      5.39 |           114.81 |            21.31 |         |
| System.Tests.Perf_String.Trim_CharArr(s: "Test ", c: [' ', ' '])                 |      5.33 |           121.05 |            22.73 |         |
| System.Tests.Perf_String.Insert(s1: "dzsdzsDDZSDZSDZSddsz", i: 7, s2: "Test")    |      5.11 |           117.97 |            23.08 |         |
| System.Tests.Perf_String.PadLeft(n: 18)                                          |      4.98 |           121.90 |            24.47 |         |
| PerfLabTests.LowLevelPerf.IntegerFormatting                                      |      4.79 |      12186752.20 |       2546510.19 |         |
| System.Tests.Perf_String.Trim_CharArr(s: " Te st  ", c: [' ', ' '])              |      4.63 |           121.91 |            26.31 |         |
| System.Tests.Perf_String.Replace_Char(text: "This is a very nice sentence", oldC |      3.36 |           139.86 |            41.60 |         |
| System.Tests.Perf_String.ToLowerInvariant(s: "TEST")                             |      3.23 |           144.91 |            44.86 |         |
| System.Tests.Perf_String.ToLowerInvariant(s: "TeSt")                             |      3.16 |           142.72 |            45.10 |         |
| System.Tests.Perf_String.ToUpperInvariant(s: "TeSt")                             |      3.16 |           140.87 |            44.63 |         |
| System.Tests.Perf_String.TrimEnd(s: "Test ")                                     |      3.15 |           145.54 |            46.21 |         |
| System.Tests.Perf_String.ToUpperInvariant(s: "test")                             |      3.10 |           138.51 |            44.67 |         |
| System.Tests.Perf_String.TrimStart(s: " Test")                                   |      3.09 |           143.10 |            46.29 |         |
| System.Tests.Perf_String.ToUpper(s: "TeSt")                                      |      2.98 |           141.85 |            47.60 |         |
| System.Tests.Perf_String.ToUpper(s: "test")                                      |      2.95 |           140.08 |            47.50 |         |
| System.Tests.Perf_String.Trim(s: " Test")                                        |      2.53 |           160.15 |            63.20 |         |
| System.Tests.Perf_String.Trim(s: "Test ")                                        |      2.53 |           160.01 |            63.28 |         |
| System.Tests.Perf_String.ToLower(s: "TEST")                                      |      2.28 |           143.54 |            62.93 |         |
| System.Tests.Perf_String.Concat_str_str(size: 100)                               |      2.25 |           172.50 |            76.50 |         |
| System.Tests.Perf_String.ToLower(s: "TeSt")                                      |      2.18 |           139.99 |            64.13 |         |
| System.Tests.Perf_String.Trim(s: " Te st  ")                                     |      2.01 |           194.77 |            96.70 |         |
| System.Tests.Perf_String.Concat_str_str_str(size: 100)                           |      1.82 |           206.83 |           113.63 |         |
| System.Tests.Perf_String.Replace_String(text: "This is a very nice sentence", ol |      1.77 |           216.99 |           122.28 |         |
| System.Tests.Perf_String.Format_MultipleArgs                                     |      1.64 |           511.59 |           311.07 |         |
| System.Tests.Perf_String.Interpolation_MultipleArgs(c: '1', s: "Foo")            |      1.64 |           512.00 |           312.67 |         |
| System.Tests.Perf_String.ToLowerInvariant(s: "This is a much longer piece of tex |      1.61 |           259.64 |           161.07 |         |
| System.Tests.Perf_String.ToUpperInvariant(s: "This is a much longer piece of tex |      1.61 |           253.11 |           157.64 |         |
| System.Tests.Perf_String.ToUpper(s: "This is a much longer piece of text that mi |      1.57 |           252.92 |           160.72 |         |
| System.Tests.Perf_String.ToLower(s: "This is a much longer piece of text that mi |      1.46 |           259.31 |           178.06 |         |
| System.Tests.Perf_String.Concat_str_str_str_str(size: 100)                       |      1.43 |           267.58 |           187.06 |         |
